Lib/typing.py

```diff
@@ -2412,10 +2414,116 @@ def utf8(value: bytes) -> bytes: ...
       def utf8(value: str) -> bytes: ...
       def utf8(value):
           # implementation goes here
+
+    The overloads for a function can be retrieved at runtime using the
+    get_overloads() function.
     """
+    # Inline version of _get_key_for_callable
+    # classmethod and staticmethod
+    inner_func = getattr(func, "__func__", func)
+    try:
+        key = (inner_func.__module__, inner_func.__qualname__)
+    except AttributeError:
+        # Not a normal function; ignore.
+        pass
+    else:
+        # We're inlining get_overloads() here to avoid computing the key twice.
+        existing = _overload_registry[key]
+        if existing and len(existing) > 20:
+            # If we are registering an overload with a lineno below or equal to that of the
+            # most recent existing overload, we're probably re-creating overloads for a
+            # function that already exists. This can happen if the module is being reloaded
+            # or if the overloads are being created in a nested function.
+            # In that case, we clear the existing overloads
+            # to avoid leaking memory. But we do this only if there are already a lot of
+            # overloads, so we don't have to figure out the linenos in the common case
+            # where there are only a few overloads.
+            try:
+                firstlineno = inner_func.__code__.co_firstlineno
+            except AttributeError:
+                pass
+            else:
+                existing_func = existing[-1]
+                # classmethod and staticmethod
+                existing_func = getattr(existing_func, "__func__", existing_func)
+                try:
+                    existing_lineno = existing_func.__code__.co_firstlineno
+                except AttributeError:
+                    pass
+                else:
+                    if firstlineno <= existing_lineno:
+                        existing.clear()
+
+        existing.append(func)
     return _overload_dummy
 
 
+# {key: [overload]}
+_overload_registry = collections.defaultdict(list)
+
+
+def get_overloads(func):
+    """Return all defined overloads for *func* as a sequence."""
+    key = _get_key_for_callable(func)
+    overloads = _overload_registry.get(key, [])
```

Comment:
I'm a little confused. Since it's a defaultdict, why not
```suggestion
    overloads = _overload_registry[key]
```
or am I a collections noob ?

Reply:
Good point, missed updating this when I changed to a defaultdict

Reply:
Using `__getitem__` here will lead to a needless entry being added to the registry if the function has 0 overloads associated with it (because it's a `defaultdict`). So `.get()` here is better in terms of memory usage.

It's probably not much of an issue, since code is unlikely to be calling `get_overloads()` repeatedly, so there's not really that much chance of a memory leak. But there is a difference here.